### PR TITLE
Increase memory to resolve file upload errors

### LINKED ISSUE #2350: Increase memory limit for deployed instances and celery workers
**Description:**

Following #2268, one reason users get error while uploading files (this happens for files > 300KB) is memory limit set on celery workers per child defined in ``` celery -A tdpservice.settings worker -c 1 --max-memory-per-child 5000 & ```.   The setting causes task to terminate when memory usage is higher than the limit (here 5MB) and consequently the child worker being replaced by another one. This means the upload task will be terminated and messages already consumed will be lost.

According to celery documentation, there are approaches to reduce the chance of such failures by adjusting the number of messages to be consumed before one task is complete, controlling how much memory is available to each child and setting when a task is deemed as complete. Despite all the above settings, upload worker still needs more memory, so all the above mentioned settings only ensures the worker is not immaturely terminated.

**Acceptance Criteria:** 

- [x] Uploaded files are scanned by clamav and only files returned as safe are uploaded to SFTP
- [x] Files are uploaded successfully to SFTP
- [x] Upload tasks are acknowledged only after completion meaning if the worker is restarted, the task will be retried
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

**Tasks:**

- [x] Increase cloud.gov instance memory (currently set to 512MB). Suggested value is 2G.
- [x] Increase per child max-memory
- [x] following settings in [here](https://docs.celeryq.dev/en/stable/userguide/optimizing.html#reserve-one-task-at-a-time) to set worker only upload one file per worker
- [x] Run Testing Checklist and confirm all tests pass

**Supporting Docs and Links**
- [Setting worker maxmemory](https://docs.celeryq.dev/en/stable/userguide/configuration.html#std-setting-worker_max_memory_per_child)
- [Celery optimization](https://docs.celeryq.dev/en/stable/userguide/optimizing.html)
- [Deployment instance config manifest](https://github.com/raft-tech/TANF-app/blob/develop/tdrs-backend/manifest.yml)
- [Example files](https://hhsgov.sharepoint.com/sites/TANFDataPortalOFA/Shared%20Documents/Forms/AllItems.aspx?csf=1&web=1&e=Jq8SRo&cid=83ce52a9%2D83f5%2D49dd%2Dac12%2Df2f7dac19ed0&RootFolder=%2Fsites%2FTANFDataPortalOFA%2FShared%20Documents%2FGeneral%2FLegacy%20System%2FDummy%20data&FolderCTID=0x01200039BDB10EB450C04FB647A089E033E116)

### PR DESCRIPTION
**Pull request by raftmsohani**
## Summary of Changes
Pull request closes #2350 _

Deployed memory increased from 512M to 2G, which resolves upload errors. Additionally, when files are missing name (see [ref](https://github.com/raft-tech/TANF-app/pull/1918#pullrequestreview-1108252379)), the log model throws error because filename is null. Changed the assigned filename to string 'None'.

## How to Test
Memory is not limited, so the error should not be happening locally. To recreate memory error locally, we should use:
```
    mem_limit: 256m
    deploy:
      resources:
        limits:
          memory: 220M
```
in docker compose.

When deployed,

1. Open http://localhost:3000/ and sign in.
1. Proceed with submitting files
1. make sure the files are uploaded to SFTP server and logs are not throwing any error.
1. Login into admin page, look at the upload logs and the files.

## Deliverables

Generally, files should be uploaded without any issue.

### [Deliverable 1: Accepted Features](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] Test files are uploaded and there is no error in the logs
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

**Comment by raftmsohani** [comment]
(Being updated as the test continues)
Assumptions:
 - An SFTP service is created in AWS EC2 instance instead of TITAN. 

The following steps were tested:

1) Increased memory to 2G from 512M, without changing the celery configuration error are happening, but they are less frequent. Having ```max-memory-per-child``` in celery config causes the worker drops because the setting tells celery whenever worker memory usage is higher than the limit, worker should be dropped to stop memory leak.
2) Decreased the memory to 512M again and reduced per child memory. Worker drops are more frequent and happen for smaller file sizes.
3) The memory issue can be recreated locally. Changes the docker settings and limited the memory assignment for wb docker. Additionally, moved redis server to web docker (this is closer to what exists in cloud.gov)
4) limiting memory for local docker web:
       a. limiting to 64: the workers constantly fail.
       b. limiting to 128: celery workers are fine but files cannot be uploaded. The code fails prefetch: this is before worker starts. looks like the code is not even getting to the upload task. Looking into logs, the files are submitted from frontend but never get to the backend because the gunicorn workers are killed as soon as the files hit the backend endpoint
There are two failures happening for mem < 210m:
                i. linux core worker is being killed
                ii. gunicorn workers are restarted.
       c. removing the memory limit: everything works as expected. So to make the memory limit similar to cloud.gov limitation, I increased the memory limit to 512 and added redis to local web.

Also added the following configuration to upload task, which tell celery to assign only one task to each worker, and tasks are deemed done only when the successful signal is back from the task.
```
task_acks_late = True
worker_prefetch_multiplier = 1
```

File: tdrs-backend/tdpservice/scheduling/sftp_task.py [hunk 1]
```diff
@@ -78,8 +78,8 @@ def create_dir(directory_name, sftp_server):
                           pkey=pkey,
                           username=username,
                           port=port,
-                          look_for_keys=False,
-                          disabled_algorithms={'pubkeys': ['rsa-sha2-512', 'rsa-sha2-256']})
```
Comment (raftmsohani) [inline tdrs-backend/tdpservice/scheduling/sftp_task.py]: This is only temporary for using AWS EC2 service. ACFTITAN server doesn't have RSA handshake and as such we should remove these settings.

File: tdrs-backend/gunicorn_start.sh [hunk 2]
```diff
@@ -19,10 +19,10 @@ python manage.py migrate
 python manage.py populate_stts
 python manage.py collectstatic --noinput
 
-celery -A tdpservice.settings worker -c 1 --max-memory-per-child 5000 &
+celery -A tdpservice.settings worker -c 1 &
 sleep 5
 # TODO: Uncomment the following line to add flower service when memory limitation is resolved
-# celery -A tdpservice.settings --broker=$REDIS_URI flower &
+celery -A tdpservice.settings --broker=$REDIS_URI flower &
```
Comment (raftmsohani) [inline tdrs-backend/gunicorn_start.sh:25]: We should be able to create an endpoint/view in django for monitoring services like: celery and redis that are not part of core web service. and then we can provide information in the admin page

File: tdrs-backend/manifest.yml [hunk 3]
```diff
@@ -1,7 +1,7 @@
 version: 1
 applications:
 - name: tdp-backend
-  memory: 512M
+  memory: 1G
```
Comment (andrew-jameson) [inline tdrs-backend/manifest.yml:4]: Not sure this change is relevant -- is this file obsolete?

**Comment by andrew-jameson** [comment]
File upload itself remains functional -- I'm not sure that the ally app is setup correctly for SFTP transfers as I saw timeouts:

```
   2023-01-24T09:16:48.39-0500 [APP/PROC/WEB/0] ERR [2023-01-24 14:16:48,376 ERROR sftp_task.py::upload:L106 :  Failed to upload ADS.E2J.NDM1.TS02 with error:[Errno 110] Connection timed out
   2023-01-24T09:16:48.40-0500 [APP/PROC/WEB/0] ERR [2023-01-24 14:16:48,376: ERROR/ForkPoolWorker-1] Failed to upload ADS.E2J.NDM1.TS02 with error:[Errno 110] Connection timed out
```

Happy to work with you on setting up ally and the other dev environments to make sure our SFTP tests remain functional.

**Review by ADPennington — APPROVED** [review]
thanks @raftmsohani! 🚀 

I submitted 4 large files (15MB each) at one time. it took about 20 seconds to get the successful submission banner. please note: @lilybl1 @reitermb

**Pull request merged**